airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DestinationV1V2Migrator.java

```diff
@@ -0,0 +1,171 @@
+     * @return whether the schema is as expected, if it exists and doesn't match then we're in an unrecoverable state.
+     */
+    default boolean doesAirbyteInternalNamespaceRawTableMatchExpectedV2Schema(DialectTableDefinition existingV2AirbyteRawTable) {
+        if (!schemaMatchesExpectation(existingV2AirbyteRawTable, RAW_TABLE_EXPECTED_V2_COLUMNS)) {
+            throw new UnexpectedSchemaException("Destination V2 Raw Table does not match expected Schema");
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * If the sync mode is a full refresh and we overwrite the table then there is no need to migrate
+     *
+     * @param syncMode            stream sync mode
+     * @param destinationSyncMode destination sync mode
+     * @return whether this is full refresh overwrite
+     */
+    default boolean isMigrationRequiredForSyncMode(final SyncMode syncMode, final DestinationSyncMode destinationSyncMode) {
+        return !(SyncMode.FULL_REFRESH.equals(syncMode) && DestinationSyncMode.OVERWRITE.equals(destinationSyncMode));
+    }
+
+    /**
+     * Checks if a valid destinations v2 raw table already exists
+     *
+     * @param streamConfig the raw table to check
+     * @return whether it exists and is in the correct format
+     */
+    default boolean doesValidV2RawTableAlreadyExist(final StreamConfig streamConfig) {
+        if (doesAirbyteInternalNamespaceExist(streamConfig)) {
+            final var existingV2Table = getTableIfExists(
+                new AirbyteStreamNameNamespacePair(streamConfig.id().rawNamespace(), streamConfig.id().rawName()));
+            return existingV2Table.isPresent() && doesAirbyteInternalNamespaceRawTableMatchExpectedV2Schema(existingV2Table.get());
+        }
+        return false;
+    }
+
+    /**
+     * Checks if a valid v1 raw table already exists
+     *
+     * @param rawTableNamePair the raw table to check
+     * @return whether it exists and is in the correct format
+     */
+    default boolean doesValidV1RawTableExist(AirbyteStreamNameNamespacePair rawTableNamePair) {
+        final var existingV1RawTable = getTableIfExists(rawTableNamePair);
+        return existingV1RawTable.isPresent() && doesV1RawTableMatchExpectedSchema(existingV1RawTable.get());
+    }
+
+    /**
+     * Checks to see if Airbyte's internal schema for destinations v2 exists
+     *
+     * @param streamConfig the stream to check
+     * @return whether the schema exists
+     */
+    boolean doesAirbyteInternalNamespaceExist(StreamConfig streamConfig);
+
+    /**
+     * Checks a Table's schema and compares it to an expected schema to make sure it matches
+     *
+     * @param existingTable the table to check
+     * @param columns       the expected schema
+     * @return whether the existing table schema matches the expectation
+     */
+    boolean schemaMatchesExpectation(DialectTableDefinition existingTable, Collection<String> columns);
```

Comment:
In this method we're referring to the schema or structure of the table rather than the schema/namespace/dataset so I think this is ok? and I don't have a strong preference either, would probably default to whatever is most widely used that isn't schema because of this naming collision.

Reply:
aha. good point. I like dataset to avoid colliding with airbyte's concept of namespace... but I'm not exactly excited to make a pr to do this find+replace :/